[FEA] Rng: expose host-rng-state in host-only API

MatthiasKohl opened the pull request:
<!--

Thank you for contributing to RAFT :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

Changes:
- Move `RngState` and `GeneratorType` to own header so they can be used in public APIs without a CUDA-compatible compiler (useful for python bindings etc.).
- Move the type of generator into the state object
- Add the `advance` method to the state object as well, allowing libraries to use `RngState` for host-only code
- Remove the `rng.hpp` header: if we want to expose the `Rng` object itself as a host-only object, the implementation/definition would have to be separated from the declaration, and this means it would be hard to keep a header-only library for these objects. Otherwise, there is simply no way that this `rng.hpp` actually fulfills its contract of a host-only header, so we remove it.

This is not the only way of creating a host-only API, but it's most likely the one which needs the least amount of changes in dependent libraries, and these libraries can basically "opt-in" to the host-only API.

@cjnolet Do you agree with this approach? I'd like to get this PR into 22.06 fairly soon, so we can finally go back to sane APIs and compile commands in cugraph-ops and cugraph.

MatthiasKohl commented:
This PR fixes #406

cjnolet commented on cpp/include/raft/random/rng_launch.cuh:
```diff
@@ -0,0 +1,50 @@
+#include "rng_state.hpp"
+
+namespace raft {
+namespace random {
```
I like the separation between `rng_state` and the actual functions which use it to generate distributions. I think we should continue to use the `rng.cuh` as the main entrypoint for the functions, though. Can we just add the contents of this file and add a deprecation message for the individual pieces which we intend to remove?

MatthiasKohl replied:
I tried to do this, but it does mean it's a little more tricky to properly ensure that we can separate the device-only API from the rest: you'll see I'm using some macro mechanism to make sure that certain parts get deprecated if included through `rng.cuh` rather than `rng_device.cuh` (which is what you should be using in the future for the device-only API).

Please do check this mechanism again because I might have missed something!

cjnolet commented on cpp/include/raft/random/rng_state.hpp:
```diff
@@ -0,0 +1,40 @@
+
+#pragma once
+
+#include "detail/rng_state.hpp"
```
The implementation details of `rng_state` are so trivial that I think we might be okay just exposing them to the public API. I could really go either way with this, though....

cjnolet reviewed (changes requested):
Overall I really like the separation of the rng state and the flattened functions because they are more consistent w/ the rest of raft's public APIs. I suggest we still use `rng.cuh` for the flattened API functions if we can.

MatthiasKohl commented on cpp/include/raft/random/detail/rng_state.hpp:
```diff
@@ -0,0 +1,61 @@
+/**
+ * The RNG state used to keep RNG state around on the host.
+ */
+struct RngState {
```
Makes sense, I'll move everything to the main header.

MatthiasKohl commented:
@cjnolet I believe I addressed your comments. I haven't fully tested all the possible deprecation messages, which are much more tricky since we want to keep the `rng.cuh` file.
But things compile at least and since there is no change to the actual code, I'm confident it works OK :)

cjnolet commented on cpp/include/raft/random/rng.hpp:
```diff
@@ -18,6 +18,20 @@
+  "raft/random/rng.hpp is deprecated and will be removed in a future relase. " \
+  "Please use raft/random/rng_state.hpp or raft/random/rng.cuh instead")
+#else
+#warning \
```
I had held off on doing a warning only because I Know this is immediately going to break a few consumers downstream who have enabled "warnings as errors". For this reason, I'm okay removing the warning tag.

MatthiasKohl replied:
Indeed the builds weren't working, so I removed this warning again

MatthiasKohl replied:
Actually, I'm using the same mechanism as in `raft/spatial/knn.hpp` now

cjnolet commented on cpp/include/raft/random/rng_device.cuh:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2018-2022, NVIDIA CORPORATION.
```
Since this is a new file, we should just use the current year in the copyright.

cjnolet commented on cpp/include/raft/random/detail/rng_device.cuh:
```diff
@@ -0,0 +1,720 @@
+/*
+ * Copyright (c) 2018-2022, NVIDIA CORPORATION.
```
Use only current year for new files

cjnolet commented on cpp/include/raft/random/detail/rng_impl.cuh:
```diff
@@ -16,906 +16,306 @@
-DI void custom_next(
-  GenType& gen, OutType* val, SamplingParams<OutType, LenType> params, LenType idx, LenType stride)
+template <typename DataT, typename WeightsT, typename IdxT = int>
+void sampleWithoutReplacement(RngState& rng_state,
```
One very small nitpick I have here is that the handle should always be the first argument for primitives that accept one.

MatthiasKohl replied:
I removed the handle here since it wasn't being used in the implementation at all (the stream is already given as another parameter).
This also keeps the API in line with other RNG APIs

cjnolet commented on cpp/include/raft/random/detail/rng_impl_deprecated.cuh:
```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2018-2022, NVIDIA CORPORATION.
```
New file should have current year

cjnolet commented on cpp/include/raft/random/rng_state.hpp:
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2018-2022, NVIDIA CORPORATION.
```
Use only current year

cjnolet reviewed (changes requested):
I found a few more trivial things in my second pass that I missed in the first pass but overall I think it looks great. Have you tried building cuml/cugraph with these changes yet just to make sure they build successfully? I don't know that we need to open full PRs to test this anymore unless there are direct changes to the public API, but it would be a good idea to check whether things like "warnings as errors" flags cause builds to fail.

MatthiasKohl commented:
@cjnolet I haven't tried the cuml/cugraph builds yet but I can do that once I've addressed your latest comments

MatthiasKohl commented:
@cjnolet I've addressed the issues you mentioned, and I've been able to build cuml. I'm currently building cugraph as well, and the files requiring RNG seem to compile fine at least.

cjnolet reviewed (approved):
LGTM. Thanks @MatthiasKohl!

cjnolet commented:
@gpucibot merge
